builtin/k8s/status_report_util.go

```diff
@@ -0,0 +1,81 @@
+package k8s
+
+import (
+	corev1 "k8s.io/api/core/v1"
+
+	sdk "github.com/hashicorp/waypoint-plugin-sdk/proto/gen"
+)
+
+// Translate a K8S container status into a Waypoint Health Status
+func containerStatusToHealth(
+	containerStatus corev1.ContainerStatus,
+) *sdk.StatusReport_Resource {
+	var resourceHealth sdk.StatusReport_Resource
+	resourceHealth.Name = containerStatus.Name
+
+	// ContainerStatus.State is a struct of possible container states. If one
+	// is set, that is the current state of the container
+	if containerStatus.State.Running != nil || containerStatus.Ready == true {
+		resourceHealth.Health = sdk.StatusReport_READY
+		resourceHealth.HealthMessage = "container is reporting running" // no message defined by k8s api
+	} else if containerStatus.State.Waiting != nil {
+		resourceHealth.Health = sdk.StatusReport_PARTIAL
+		resourceHealth.HealthMessage = containerStatus.State.Waiting.Message
+	} else if containerStatus.State.Terminated != nil {
+		resourceHealth.Health = sdk.StatusReport_DOWN
+		resourceHealth.HealthMessage = containerStatus.State.Terminated.Message
+	} else {
+		resourceHealth.Health = sdk.StatusReport_UNKNOWN
+		resourceHealth.HealthMessage = "container health could not be determined"
+	}
+
+	return &resourceHealth
+}
+
+// Translate a Pod Phase into a Waypoint Health Status
+func podPhaseToHealth(
+	phase corev1.PodPhase,
+) sdk.StatusReport_Health {
+	var healthResult sdk.StatusReport_Health
+
+	switch phase {
```

Comment:
I think that the pod conditions have a closer mapping our statuses than the pod phase. For example, the `PodReady` condition is documented as indicating the pod has passed health checks and can take requests, but the `Running` phase only implies that the pod has been bound to a node and the containers have been created.

https://kubernetes.io/docs/concepts/workloads/pods/pod-lifecycle/#pod-phase
https://kubernetes.io/docs/concepts/workloads/pods/pod-lifecycle/#pod-conditions

Reply:
That makes sense! I'll base the report on that for pods instead. Agreed ConditionStatus looks a lot better.

Reply:
Hrm, actually looking at this, Pod Conditions don't really describe any failure states... I see "Ready", "Initialized", "PodScheduled", and "ContainersReady" as [possible conditions](https://pkg.go.dev/k8s.io/api/core/v1#PodConditionType), where as the pod phase describes the full potential state a pod could be in: https://pkg.go.dev/k8s.io/api/core/v1#PodPhase

Reply:
Ok, got that fixed up now!